# python fundamentals

## 1.variables

### What is variable?
variable is name used to store the value in apython program

variables make it easier to store ,access and modify.


### why do use variables
variables are used to store informations such as name, age, numbers and others 

### Real world example
variable can be compared to labeled box. the label tells us that what is stored inside the box.
Ex. a box labeled name can store the name "Nandini"

In [1]:
name = "Nandini"
age = "23"

print(name)
print(age)

Nandini
23


### code explation
In the above program:
'name' is a variable thar stores the valve "Nandini"
'age' is a variable thar stores the valve "23"
print() is a function is used to display the values stored in a variables

## 2. Data types


## what is data type 
a data type tells us that what kind of valve is stored in variable

### common python data types

- data types | description
- int   | whole number
- float | decimal values
- str   | text
- bool  | true or false
- list  | ordered and changable collection
- tuple | ordered and unchangable collection
- set   | unordered and changable collection

## Why data types are important 
data types helps python to understand what kind of data we are working with

### real world example

consider a student:

- Name -> text -> 'str'
- age -> number -> 'int'
- percentage -> decimal number -> 'float'
- Passed -> yes or no -> 'bool'


In [2]:
name = "Nandini"
age = 23
percentage = 85.5
is_student = True

print(name)
print(age)
print(percentage)
print(is_student)

Nandini
23
85.5
True


### Code Explanation

- `name` contains text, so its data type is `str`.
- `age` contains a whole number, so its data type is `int`.
- `percentage` contains a decimal number, so its data type is `float`.
- `is_student` contains a Boolean value, so its data type is `bool`.
- The `type()` function is used to identify the data type of a variable.

# 3. Type Casting

## What is Type Casting?
Type casting is the process of converting a value from one data type to another data type.


### Why do we use Type Casting?
Type casting is useful when we need to work with data in a different format.


## Real-World Example

Suppose a user enters their age as "23". Python treats it as a string. If we want to add or compare the age as a number, we can convert it into an integer using `int()`.


## Common Type Casting Functions

- int() → converts a value to an integer
- float() → converts a value to a float
- str() → converts a value to a string
- bool() → converts a value to a Boolean value

In [26]:
age = "23"

print(age)
print(type(age))

age = int(age)

print(age)
print(type(age))

23
<class 'str'>
23
<class 'int'>


### Code Explanation

- Initially, "age" contains "23", which is a string.
- The type() function shows that the value is of type str.
- The int() function converts the string "23" into the integer 23.
- After conversion, the data type of age becomes int.

## 1. Arithmetic Operators

Arithmetic operators are used to perform mathematical calculations.

The main arithmetic operators in Python are:

- `+` Addition
- `-` Subtraction
- `*` Multiplication
- `/` Division
- `//` Floor Division
- `%` Modulus
- `**` Exponentiation

In [27]:
a = 10
b = 3

print("Addition:", a + b)
print("Subtraction:", a - b)
print("Multiplication:", a * b)
print("Division:", a / b)
print("Floor Division:", a // b)
print("Modulus:", a % b)
print("Exponent:", a ** b)

Addition: 13
Subtraction: 7
Multiplication: 30
Division: 3.3333333333333335
Floor Division: 3
Modulus: 1
Exponent: 1000


### Code Explanation

- `+` adds two numbers.
- `-` subtracts one number from another.
- `*` multiplies two numbers.
- `/` performs normal division.
- `//` gives the floor division result.
- `%` gives the remainder.
- `**` calculates the power of a number.

## 2. Comparison Operators

Comparison operators are used to compare two values.

The result of a comparison is a Boolean value: 'True' or 'False'

- == Equal to
- != Not equal to
- >	Greater than
- <	Less than
- >= Greater than or equal to
- <= Less than or equal to

In [28]:
a = 10
b = 5

print(a == b)
print(a != b)
print(a > b)
print(a < b)
print(a >= b)
print(a <= b)

False
True
True
False
True
False


### Code Explanation

`a == b` checks whether both values are equal.

`a != b` checks whether the values are different.

`a > b` checks whether `a` is greater than `b`.

`a < b` checks whether `a` is smaller than `b`.

`a >= b` checks whether `a` is greater than or equal to `b`.

`a <= b` checks whether `a` is less than or equal to `b`.

## 3.Assignment Operators

Assignment operators are used to assign or update values.

 Operator |	Example   |Meaning
- =	      |  x = 10	  |  Assign
- +=	  |  x += 5	  |  Add and assign
- -=      |  x -= 5	  |  Subtract and assign
- *=	  |  x *= 5	  |  Multiply and assign
- /=      |	 x /= 5	  | Divide and assign
- //=	  |  x //= 5  | Floor divide and assign
- %=	  |  x %= 5	  |  Modulus and assign
- **= 	  |  x **= 2  |    Power and assign

## 4.Logical Operators

Logical operators are used to combine conditions
There are three:

- and
- or
- not

- and

Both conditions must be true.

In [29]:
age = 23
has_id = True

print(age >= 18 and has_id)

True


- or

At least one condition should be true.

In [30]:
age = 17
has_permission = True

print(age >= 18 or has_permission)

True


- not

It reverses the Boolean result.

In [31]:
is_raining = False

print(not is_raining)

True


## 4. Logical Operators

Logical operators are used to combine or reverse conditions.

- `and` returns True when both conditions are True.
- `or` returns True when at least one condition is True.
- `not` reverses the result of a condition.

## 5.Bitwise Operators

These operators work at the binary/bit level.
The main ones are:

- &
- |
- ^
- ~
- << - >>

In [32]:
a = 5
b = 3

print(a & b)
print(a | b)
print(a ^ b)

1
7
6


### Code Explanation

In this program, two integer variables a and b are created with the values 5 and 3.

- a = 5
- b = 3

The numbers are converted into binary form:

- 5 → 101
- 3 → 011

# 5. Input and Output
## What is input?
Input means taking information from the user
- Python provides the `input()` function to take input from the user.



In [3]:
name = input("Enter your name: ")

print("Hello", name)

Enter your name:  Nandini Hasabi


Hello Nandini Hasabi


## What is Output?
Output means displaying information or results to the user.
- In Python, the print() function is used to display output on the screen.

In [21]:
name = "Nandini"
age = 23

print("Name:", name)
print("Age:", age)

Name: Nandini
Age: 23


## Real-World Example

A registration form asks a user for their name, age, and city.

The program takes these details as input and displays them back to the user.

This demonstrates how `input()` is used to receive information and `print()` is used to display information.

In [22]:
name = input("Enter your name: ")
age = int(input("Enter your age: "))
city = input("Enter your city: ")

print("\nRegistration Details")
print("Name:", name)
print("Age:", age)
print("City:", city)

Enter your name:  nandini
Enter your age:  23
Enter your city:  bangalore



Registration Details
Name: nandini
Age: 23
City: bangalore


### Code Explanation

- input() is used to receive information from the user.
- name and city are stored as strings.
- int() converts the age entered by the user from a string into an integer.
- print() is used to display the entered information.
- The \n creates a new line before displaying the registration details.

# 6. Comments

## What is a Comment?

A comment is a line of text written in a Python program to explain the code.

Python ignores comments when the program is executed.

## Why do we use Comments?

Comments can be used to:

- Explain what a part of the code does.
- Make complex code easier to understand.

Python mainly uses '#' for single-line comments.

## 1. Single-Line Comment

Use # to create a comment.

In [23]:
# Store the user's name
name = "Nandini"

# Display the name
print(name)

Nandini


### Code Explanation

- "# Store the user's name" is a comment that explains what the next line does.
- 'name = "Nandini"' stores the name in a variable.
- '# Display the name' explains the purpose of the next line.
- 'print(name)' displays the value stored in the variable.
- Python ignores the lines beginning with '#'.

## 2.Multiple-Line Comments

Python doesn't have a special multiple-line comment symbol like some other programming languages.

You can use # on each line:

In [25]:
# This program stores
# a student's name
# and displays it.

name = "Nandini"
print(name)

Nandini


# 7. Control Flow

Control flow means controlling the order in which statements in a Python program are executed.

Normally, Python executes statements from top to bottom. Control flow statements allow us to make decisions, repeat a block of code, or change the normal execution of a program.

The main control flow statements in Python are:
- if
- elif
- else
- for loop
- while loop
- break
- continue
- pass

### Example

Imagine a traffic signal. If the signal is green, we can go. If it is red, we must stop. The action depends on a condition.

Similarly, Python uses control flow statements to decide which code should execute.

In [4]:
age = 20

if age >= 18:
    print("You are eligible to vote")
else:
    print("You are not eligible to vote")
    

You are eligible to vote


### Code Explanation

The variable 'age' stores the value 20.

The 'if' statement checks whether the age is greater than or equal to 18.

Since the condition is true, Python executes the 'print()' statement inside the `if` block.

If the condition were false, the 'else' block would execute.

## if Statement

The 'if' statement is used to execute a block of code only when a given condition is true.

If the condition is true, the code inside the 'if' block is executed. If the condition is false, it is skipped.

In [5]:
temperature = 30

if temperature > 25:
    print("It is a hot day")

It is a hot day


### Code Explanation

'temperature' stores the value 30.

The 'if' statement checks whether the temperature is greater than 25.

Since 30 is greater than 25, the message "It is a hot day" is printed.

## if-else Statement

The 'if-else' statement is used when we have two possible outcomes.

If the condition is true, the 'if' block is executed. Otherwise, the 'else' block is executed.

In [6]:
number = 7

if number % 2 == 0:
    print("Even number")
else:
    print("Odd number")

Odd number


### Code Explanation

The '%' operator gives the remainder after division.

The program checks whether the remainder when 'number' is divided by 2 is zero.

If it is zero, the number is even. Otherwise, the number is odd.

## if-elif-else Statement

The 'if-elif-else' statement is used when we need to check multiple conditions.

Python first checks the 'if' condition. If it is false, it checks the 'elif' condition. If all conditions are false, the 'else' block is executed.

In [7]:
marks = 75

if marks >= 90:
    grade = "A"
elif marks >= 75:
    grade = "B"
elif marks >= 50:
    grade = "C"
else:
    grade = "Fail"

print("Grade:", grade)

Grade: B


### Code Explanation

The variable 'marks' stores 75.

Python first checks whether the marks are greater than or equal to 90. This condition is false.

Then it checks whether the marks are greater than or equal to 75. This condition is true, so 'grade' is assigned "B".

Python does not check the remaining conditions after finding a true condition.

## for Loop

A 'for' loop is used to repeat a block of code for each item in a sequence or collection.

It is useful when we know what items we want to iterate through.

In [8]:
fruits = ["Apple", "Banana", "Mango"]

for fruit in fruits:
    print(fruit)

Apple
Banana
Mango


### Code Explanation

The list 'fruits' contains three items.

The 'for' loop takes one item at a time from the list and stores it in the variable 'fruit'.

The 'print()' statement displays each fruit.

## while Loop

A 'while' loop repeatedly executes a block of code as long as a given condition is true.

It is useful when the number of repetitions depends on a condition.

In [9]:
count = 1

while count <= 5:
    print(count)
    count += 1

1
2
3
4
5


### Code Explanation

The variable 'count' starts with the value 1.

The 'while' loop checks whether 'count' is less than or equal to 5.

The value is printed and then increased by 1 using 'count += 1'.

The loop stops when 'count' becomes 6 because the condition becomes false.

## break Statement

The 'break' statement is used to immediately stop a loop.

When Python reaches 'break', it exits the loop even if the loop condition is still true.

In [10]:
for number in range(1, 10):
    if number == 5:
        break
    print(number)

1
2
3
4


### Code Explanation

The loop starts from 1 and goes up to 9.

When 'number' becomes 5, the 'if' condition becomes true.

The 'break' statement immediately stops the loop.

Therefore, the numbers printed are 1, 2, 3, and 4.

## continue Statement

The 'continue' statement skips the current iteration of a loop and moves to the next iteration.

It does not stop the complete loop.

In [11]:
for number in range(1, 6):
    if number == 3:
        continue
    print(number)

1
2
4
5


### Code Explanation

The loop runs from 1 to 5.

When 'number' becomes 3, 'continue' skips that iteration.

The loop then continues with the next number.

Therefore, the output is:

1
2
4
5

## pass Statement

The 'pass' statement is a placeholder that does nothing when executed.

It is useful when we want to create a block of code but do not want to write its implementation yet.

In [12]:
for number in range(1, 6):
    if number == 3:
        pass
    else:
        print(number)

1
2
4
5


### Code Explanation

When 'number' is 3, the 'pass' statement is executed.

'pass' does not perform any action, so nothing happens for that iteration.

The loop continues normally.

## Collections

Collections are used to store and organize multiple values in a single variable.

Python provides different collection types for storing data in different ways.

The main collection types covered in this sprint are:

- List
- Tuple
- Set
- Dictionary

Each collection has different properties and is useful for different situations.

### Example

Collections are like different types of containers.

A list is like a shopping bag where we can add, remove, or change items.

A tuple is like a sealed box where the items cannot be changed after creation.

A set is like a collection of unique items where duplicates are removed.

A dictionary is like a real-world dictionary where each word has a corresponding meaning. In Python, it stores data as key-value pairs.

## 1. List

A list is an ordered and mutable collection used to store multiple values.

Lists can contain duplicate values and different types of data.

Lists are created using square brackets '[]'.

In [13]:
fruits = ["Apple", "Banana", "Mango", "Apple"]

print(fruits)

['Apple', 'Banana', 'Mango', 'Apple']


## 2. Tuple

A tuple is an ordered collection used to store multiple values.

Tuples are immutable, which means their elements cannot be changed after creation.

Tuples are created using parentheses '()'.

In [14]:
coordinates = (10, 20, 30)

print(coordinates)

(10, 20, 30)


## 3.Set

A set is a collection used to store unique values.

Sets do not allow duplicate elements.

Sets are mutable, so we can add or remove elements.

Sets are created using curly brackets '{}'.

## 4.Dictionary

A dictionary is a collection that stores data in key-value pairs.

Each key is used to access its corresponding value.

Dictionaries are mutable.

Dictionaries are created using curly brackets '{}' with key-value pairs.

## List Comprehension

List comprehension is a simple and shorter way to create a new list from an existing sequence.

It allows us to write a loop and optionally a condition in a single line.

The basic syntax is:

[expression for item in iterable]

In [15]:
numbers = [1, 2, 3, 4, 5]

squares = [number * number for number in numbers]

print(squares)

[1, 4, 9, 16, 25]


### Code Explanation

The list 'numbers' contains five numbers.

The list comprehension takes each number from the 'numbers' list and multiplies it by itself.

The expression 'number * number' calculates the square of each number.

The resulting values are stored in a new list called 'squares'.

List comprehension makes the code shorter compared to using a traditional 'for' loop.

## Dictionary Comprehension

Dictionary comprehension is a shorter and simpler way to create a dictionary using an iterable.

It allows us to create key-value pairs using a single line of code.

The basic syntax is:

{key: value for item in iterable}

### Example

Suppose we have a list of numbers:

[1, 2, 3, 4, 5]

We want to create a dictionary where each number is a key and its square is the value.

For example:

- 1 → 1
- 2 → 4
- 3 → 9
- 4 → 16
- 5 → 25

Dictionary comprehension allows us to create this dictionary in a short and readable way.

In [16]:
numbers = [1, 2, 3, 4, 5]

squares = {number: number * number for number in numbers}

print(squares)

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25}


### Code Explanation

The list 'numbers' contains five numbers.

The dictionary comprehension takes each number from the list.

The number is used as the key, and 'number * number' is used as the value.

The resulting key-value pairs are stored in the 'squares' dictionary.

Dictionary comprehension makes the code shorter and easier to read compared to a traditional `for` loop.

## Function Definition

A function is a reusable block of code that performs a specific task.

Instead of writing the same code multiple times, we can define a function once and call it whenever we need it.

In Python, a function is defined using the 'def' keyword.

### Example

A function is like a machine.

We give the machine an input, it performs a task, and it can give us an output.

For example, a calculator can perform addition whenever we provide two numbers.

In [17]:
def greet():
    print("Hello, welcome to Python!")

greet()

Hello, welcome to Python!


### Code Explanation

The 'def' keyword is used to define a function.

'greet' is the name of the function.

The code inside the function prints a welcome message.

The statement 'greet()' calls the function and executes the code inside it.

## Function Arguments

An argument is a value passed to a function when the function is called.

Arguments allow the same function to work with different values.

In [18]:
def greet(name):
    print("Hello", name)

greet("Nandini")
greet("Rahul")

Hello Nandini
Hello Rahul


### Code Explanation

`name` is the parameter of the function.

When we call `greet("Nandini")`, the value `"Nandini"` is passed to the parameter `name`.

When we call `greet("Rahul")`, `"Rahul"` is passed to `name`.

This allows us to reuse the same function with different values.

## Function with Return Value

In [19]:
def add(a, b):
    result = a + b
    return result

answer = add(10, 20)

print(answer)

30


### Code Explanation

The function 'add()' takes two arguments, 'a' and 'b'.

The values are added and stored in 'result'.

The 'return' statement sends the result back to the place where the function was called.

The returned value is stored in the variable 'answer'.

## Default Arguments
A default argument is a value that is already given to a parameter.

If we do not provide a value when calling the function, Python uses the default value.

Example
- If no name is given, we can use "Guest" as the default name.



In [21]:
def greet(name="Guest"): 
    print("Hello", name)
    greet()
    greet("Nandini")

### Explanation
- name="Guest" gives a default value to the parameter.
- When greet() is called without an argument, "Guest" is used.
- When "Nandini" is provided, Python uses "Nandini" instead of the default value.

## 4. Keyword Arguments

Keyword arguments allow us to pass values by using the parameter names.

This makes the function call easier to understand.

### Example

We can pass a student's name and age using their parameter names.

In [22]:
def student_info(name, age):
    print("Name:", name)
    print("Age:", age)

student_info(name="Nandini", age=23)

Name: Nandini
Age: 23


### Explanation

- 'name' and 'age' are parameters.
- 'name="Nandini"' and 'age=23' are keyword arguments.
- We clearly specify which value belongs to which parameter.

## Variable-Length Arguments (*args)

Sometimes we do not know how many arguments will be passed to a function.

`*args` allows a function to accept multiple positional arguments.

### Example

We can create a function that adds any number of numbers.

In [23]:
def add_numbers(*numbers):
    total = 0

    for number in numbers:
        total = total + number

    print("Total:", total)

add_numbers(10, 20)
add_numbers(10, 20, 30, 40)

Total: 30
Total: 100


### Explanation

- '*numbers' allows us to pass multiple values.
- The values are stored together.
- The 'for' loop takes each number one by one.
- The numbers are added and the total is printed.

'*args' is useful when we do not know the number of positional arguments in advance.

## 6. Variable-Length Keyword Arguments (**kwargs)

'**kwargs' allows a function to accept multiple keyword arguments.

The values are stored as a dictionary.

### Example

We can use '**kwargs' to store different student details.

In [24]:
def student_details(**details):
    print(details)

student_details(name="Nandini", age=23, city="Bangalore")

{'name': 'Nandini', 'age': 23, 'city': 'Bangalore'}


### Explanation

- '**details' accepts multiple keyword arguments.
- The arguments are stored as a dictionary.
- 'name', 'age', and 'city are keys.
- Their values are stored with the corresponding keys.

### Remember

*args → multiple positional arguments

**kwargs → multiple keyword arguments

## 7. Lambda Functions

A lambda function is a small function written in a single line.

It is useful when we need a simple function for a short task.

### Example

We can use a lambda function to find the square of a number.

In [25]:
square = lambda x: x * x

print(square(5))

25


### Explanation

- 'lambda' is used to create a small function.
- 'x' is the input.
- 'x * x' calculates the square.
- 'square(5)' returns '25'.

## 8. Recursive Functions

A recursive function is a function that calls itself.

A recursive function must have a stopping condition. This is called the base condition.

### Example

We can use recursion to calculate the factorial of a number.

In [26]:
def factorial(number):
    if number == 0:
        return 1

    return number * factorial(number - 1)

print(factorial(5))

120


### Explanation

- 'factorial()' is a recursive function because it calls itself.
- 'if number == 0' is the base condition.
- The function keeps calling itself with a smaller number.
- When the number becomes '0', the function stops.
- The factorial of 5 is '120'.

Calculation:

5 × 4 × 3 × 2 × 1 = 120

# Advanced Python

Advanced Python concepts help us write more flexible, reusable, and efficient Python programs.

In this section, I will learn about:

- Iterators
- Generators
- yield
- Decorators
- Closures
- global
- nonlocal

## 1. Iterators

An iterator is an object that gives us one value at a time.

We can use an iterator to go through the elements of a collection one by one.

Python provides two important functions for working with iterators:

- 'iter()' → creates an iterator
- 'next()' → gets the next value from the iterator

### Example

We can create an iterator from a list and get each value one by one.

In [27]:
numbers = [10, 20, 30]

numbers_iterator = iter(numbers)

print(next(numbers_iterator))
print(next(numbers_iterator))
print(next(numbers_iterator))

10
20
30


### Explanation

- 'numbers' is a list containing three values.
- 'iter()' converts the list into an iterator.
- 'next()' gives the next value from the iterator.
- Each time we call 'next()', we get the next value.

The values are returned one at a time instead of getting all values at once.

## 2. Generators

A generator is a special type of function that produces values one at a time.

Generators are useful when we have a large amount of data because they do not store all the values in memory at the same time.

A generator function uses the `yield` keyword.

### Example

We can create a generator that produces numbers from 1 to 3.

In [28]:
def numbers():
    yield 1
    yield 2
    yield 3

for number in numbers():
    print(number)

1
2
3


### Explanation

- 'numbers()' is a generator function.
- 'yield' produces one value at a time.
- The 'for' loop gets each value from the generator.
- The generator does not create all the values at the same time.

Generators are useful when working with large amounts of data.

## 3. yield

The 'yield' keyword is used in a generator function.

Unlike 'return', 'yield' does not completely stop the function. It pauses the function and remembers its current state.

When the function is called again, it continues from where it stopped.


In [29]:
def count_numbers():
    yield 1
    yield 2
    yield 3

result = count_numbers()

print(next(result))
print(next(result))
print(next(result))

1
2
3


### Explanation

- 'yield 1' gives the first value.
- The function pauses after producing the value.
- The next 'next()' call continues the function.
- It then produces '2' and then '3'.

The main difference is:

'return' → returns a value and ends the function.

'yield' → returns a value and pauses the function.

## 4. Decorators

A decorator is used to add extra functionality to an existing function without changing the original function's code.

A decorator takes a function and adds some additional behavior to it.

### Example

We can create a decorator that prints a message before calling a function.

In [30]:
def my_decorator(function):
    def wrapper():
        print("Starting the function")
        function()
    
    return wrapper


@my_decorator
def greet():
    print("Hello!")

greet()

Starting the function
Hello!


### Explanation

- 'my_decorator' is a decorator function.
- It receives another function as input.
- 'wrapper()' adds extra behavior.
- '@my_decorator' applies the decorator to the 'greet()' function.
- When 'greet()' is called, the decorator runs first and then the original function runs.

Decorators allow us to add functionality without changing the original function.

## 5. Closures

A closure happens when an inner function remembers a value from the outer function even after the outer function has finished running.

In simple words, the inner function can remember information from its surrounding function.

### Example

We can create a function that remembers a person's name.

In [31]:
def create_greeting(name):
    def greet():
        print("Hello", name)
    
    return greet


greeting = create_greeting("Nandini")

greeting()

Hello Nandini


### Explanation

- 'create_greeting()' is the outer function.
- 'greet()' is the inner function.
- The inner function uses the 'name' variable from the outer function.
- Even after 'create_greeting()' finishes, 'greet()' remembers the value of 'name'.
- This is called a closure.

## 6. global

The 'global' keyword is used when we want to change a variable that is created outside a function.

A variable created outside a function is called a global variable.

In [32]:
count = 10

def update_count():
    global count
    count = 20

update_count()

print(count)

20


### Explanation

- 'count' is created outside the function, so it is a global variable.
- Normally, a function cannot directly change the global variable.
- The 'global' keyword tells Python that we want to use the global variable.
- The value of 'count' is changed from '10' to '20'.

## 7. nonlocal
The `nonlocal` keyword is used inside a nested function when we want to change a variable from the outer function.

It is mainly used with nested functions and closures.


In [33]:
def counter():
    count = 0

    def increase():
        nonlocal count
        count = count + 1
        print(count)

    increase()
    increase()

counter()

1
2


### Explanation

- 'count' belongs to the outer 'counter()' function.
- 'increase()' is an inner function.
- 'nonlocal' allows the inner function to change the 'count' variable.
- Each time 'increase()' is called, the value of 'count' increases.
- The output is '1' and '2'.

The 'nonlocal' keyword is used for variables in an enclosing function.

## 1. Modules

A module is a Python file that contains code such as functions, variables, and classes.

We can use the code from a module in another Python program by importing it.

Modules help us organize our code and reuse it in different programs.

### Example

Python already provides many built-in modules. We can import them and use their functions.

In [34]:
import math

number = 25

print(math.sqrt(number))

5.0


### Explanation

- 'import math' imports the built-in 'math' module.
- 'math.sqrt()' is used to find the square root of a number.
- Here, the square root of '25' is '5.0'.
- Instead of writing our own square root calculation, we can use the function provided by the module.

Modules make code easier to reuse and organize.

## Why Do We Use Modules?

Modules are useful because they:

- Help us reuse code.
- Help us organize code.
- Reduce repeated code.
- Make programs easier to maintain.
- Provide ready-made functions that we can use.

## Types of Modules

There are two common types of modules:

1. Built-in Modules
2. User-defined Modules

### Built-in Modules

Built-in modules are already provided by Python.

Some examples are:

- `math` - used for mathematical operations
- `random` - used to generate random values
- `datetime` - used to work with dates and times
- `os` - used to work with the operating system

In [35]:
import random

number = random.randint(1, 10)

print("Random number:", number)

Random number: 1


### Explanation

- 'import random' imports the 'random' module.
- 'random.randint(1, 10)' generates a random integer between 1 and 10.
- The result can be different each time we run the program.

## User-Defined Module

A user-defined module is a Python file created by the programmer.

For example, we can create a file called `calculator.py` and add our own functions to it.

In [36]:
# calculator.py

def add(a, b):
    return a + b

def multiply(a, b):
    return a * b

### Using the User-Defined Module

After creating `calculator.py`, we can import it into another Python program and use its functions.

In [37]:
import calculator

print(calculator.add(10, 20))
print(calculator.multiply(5, 4))

ModuleNotFoundError: No module named 'calculator'

### Explanation

- 'calculator.py' is our user-defined module.
- It contains the 'add()' and 'multiply()' functions.
- 'import calculator' imports the module.
- 'calculator.add()' calls the 'add()' function from the module.
- 'calculator.multiply()' calls the 'multiply()' function from the module.

This allows us to reuse functions without writing them again.

## Importing a Specific Function

We can import only the function that we need from a module.

This can make the code shorter and easier to read.

In [38]:
from math import sqrt

print(sqrt(36))

6.0


### Explanation

- 'from math import sqrt' imports only the 'sqrt()' function from the 'math' module.
- We can use 'sqrt()' directly without writing 'math.sqrt()'.
- The output is '6.0'.

## Using an Alias

We can give a shorter name to a module using the `as` keyword.

This shorter name is called an alias.

In [39]:
import math as m

print(m.sqrt(49))

7.0


### Explanation

- 'math' is imported with the alias 'm'.
- We can use 'm' instead of writing 'math'.
- 'm.sqrt(49)' finds the square root of 49.
- The output is '7.0'.

## What is a Package?

A package is a collection of related Python modules stored in a folder.

Packages help us organize multiple modules in a structured way.

For example, if a project has many Python files, we can group related files inside a package.

## Example

Suppose we have a project for a calculator.

We can organize it like this:

calculator_package/
    addition.py
    subtraction.py
    multiplication.py

Here, 'calculator_package' is the package and the Python files inside it are modules.

A package helps us keep related modules together.

## Creating a Package

A simple package can contain multiple Python files.

For example:

calculator_package/
    __init__.py
    addition.py
    multiplication.py

The '__init__.py' file tells Python that the folder can be treated as a package.

In modern Python versions, some packages can also work without '__init__.py', but it is commonly used for regular packages.

In [40]:
# addition.py

def add(a, b):
    return a + b

## Importing Modules from a Package

We can import a module from a package and use its functions.

In [41]:
from calculator_package import addition

print(addition.add(10, 20))

ModuleNotFoundError: No module named 'calculator_package'

### Explanation

- 'calculator_package' is the package.
- 'addition' is a module inside the package.
- 'add()' is a function inside the 'addition' module.
- We import the 'addition' module from the package.
- Then we use 'addition.add()' to call the function.

## Importing a Function Directly

We can also import a specific function from a module inside a package.

In [42]:
from calculator_package.addition import add

print(add(5, 10))

ModuleNotFoundError: No module named 'calculator_package'

### Explanation

Here, we directly import the 'add()' function from the 'addition' module.

This allows us to call 'add()' directly without writing 'addition.add()'.

## Why Do We Use Packages?

Packages are useful because they:

- Organize related modules.
- Make large projects easier to manage.
- Help us reuse code.
- Make the project structure cleaner.
- Avoid keeping all code in one large Python file.

# Virtual Environments

## What is a Virtual Environment?

A virtual environment is an isolated environment used to install and manage Python packages for a specific project.

It keeps the packages of one project separate from the packages of other projects.

For example, one project may require TensorFlow 2.21, while another project may require a different version. Virtual environments help avoid conflicts between them.

## Why Do We Use Virtual Environments?

Virtual environments are useful because they:

- Keep project dependencies separate.
- Avoid conflicts between different package versions.
- Make projects easier to manage.
- Allow each project to have its own required packages.
- Help create a consistent development environment.

## Creating a Virtual Environment

We can create a virtual environment using the `venv` module provided by Python.

The following command creates a virtual environment named `myenv`.

In [ ]:
python -m venv myenv

## Activating a Virtual Environment

After creating the virtual environment, we need to activate it before installing packages.

### Windows

Use the following command:

myenv\Scripts\activate

### Linux or macOS

Use the following command:

'source myenv/bin/activate'

After activation, packages installed using 'pip' will be installed inside the virtual environment.

## Installing a Package

After activating the virtual environment, we can install Python packages using `pip`.

For example, we can install NumPy.

In [43]:
pip install numpy

Note: you may need to restart the kernel to use updated packages.


## Checking Installed Packages

We can use the following command to see the packages installed in the current environment.

In [ ]:
pip list

## Deactivating a Virtual Environment

After finishing our work, we can deactivate the virtual environment using:

In [ ]:
deactivate

## Example

Suppose I have two Python projects:

- Project A requires NumPy version 1.x.
- Project B requires a different NumPy version.

Installing both versions in the same environment can cause conflicts.

Instead, I can create two separate virtual environments:

Project A → Environment A → Required packages

Project B → Environment B → Required packages

This keeps the dependencies of both projects separate.

# Type Hinting 

## What is Type Hinting?

Type hinting is a way of specifying the expected data type of a variable, function parameter, or function return value.

It makes the code easier to understand and helps developers identify the expected type of data.

Python is still dynamically typed, so type hints do not force the type at runtime.

## Type Hinting for Variables

We can specify the expected type of a variable using a colon ':'.

### Example

Here, 'name' is expected to contain a string and 'age' is expected to contain an integer.

In [44]:
name: str = "Nandini"
age: int = 23

print(name)
print(age)

Nandini
23


### Explanation

- 'str' means the variable is expected to contain a string.
- 'int' means the variable is expected to contain an integer.
- Type hints make the expected data type clear to other developers.

## Type Hinting with Lists

We can also specify the type of elements expected inside a list.

For example, `list[int]` means a list containing integers.

In [45]:
numbers: list[int] = [10, 20, 30, 40]

print(numbers)

[10, 20, 30, 40]


# Built-in Functions in Python

## What are Built-in Functions?

Built-in functions are functions that are already provided by Python.

We can use them directly without defining them ourselves.

Some commonly used built-in functions are:

- 'map()'
- 'filter()'
- 'zip()'
- 'enumerate()'
- 'sorted()'
- 'any()'
- 'all()'
- 'isinstance()'
- 'id()'

These functions help us perform common operations easily and efficiently.

## 1. map()

The 'map()' function applies a function to every item in an iterable such as a list.

It is useful when we want to perform the same operation on multiple values.

### Example

We can use 'map()' to find the square of every number in a list.

In [46]:
numbers = [1, 2, 3, 4, 5]

squares = list(map(lambda x: x * x, numbers))

print(squares)

[1, 4, 9, 16, 25]


### Explanation

- 'numbers' contains the values from 1 to 5.
- 'map()' applies the lambda function to every number.
- The lambda function calculates the square.
- 'list()' converts the result into a list.
- The output contains the squares of all the numbers.

## 2. filter()

### Concept

The 'filter()' function selects items from an iterable based on a condition.

It keeps only the values for which the condition is 'True'.

### Example

We can use 'filter()' to get only even numbers from a list.

In [47]:
numbers = [1, 2, 3, 4, 5, 6]

even_numbers = list(filter(lambda x: x % 2 == 0, numbers))

print(even_numbers)

[2, 4, 6]


### Explanation

- 'filter()' checks each number using the given condition.
- 'x % 2 == 0' checks whether the number is even.
- Only numbers that satisfy the condition are selected.
- The output is '[2, 4, 6]'.

## 3. zip()

The `zip()` function combines elements from two or more iterables.

It pairs the elements based on their positions.

### Example

We can combine a list of names with a list of ages.

In [48]:
names = ["Nandini", "Anu", "Priya"]
ages = [23, 22, 24]

students = list(zip(names, ages))

print(students)

[('Nandini', 23), ('Anu', 22), ('Priya', 24)]


### Explanation

- `zip()` combines the elements from both lists.
- The first name is paired with the first age.
- The second name is paired with the second age.
- The result contains pairs in the form of tuples.

## 4. enumerate()

The `enumerate()` function gives us both the index and the value while looping through an iterable.

It is useful when we need to know the position of each item.


In [49]:
fruits = ["Apple", "Banana", "Mango"]

for index, fruit in enumerate(fruits):
    print(index, fruit)

0 Apple
1 Banana
2 Mango


### Explanation

- `enumerate()` provides both the index and the value.
- `index` stores the position.
- `fruit` stores the current item.
- By default, indexing starts from `0`.

Output:

0 Apple
1 Banana
2 Mango

## 5. sorted()

The `sorted()` function returns a new sorted list from an iterable.

By default, it sorts values in ascending order.

### Example

In [50]:
numbers = [40, 10, 30, 20]

sorted_numbers = sorted(numbers)

print(sorted_numbers)

[10, 20, 30, 40]


### Explanation

- `sorted()` arranges the values in ascending order.
- The original list is not changed.
- The result is a new sorted list.

Output:

[10, 20, 30, 40]

## 6. any()

The `any()` function returns `True` if at least one value in an iterable is `True`.

If all values are `False`, it returns `False`.


In [51]:
values = [False, False, True, False]

result = any(values)

print(result)

True


### Explanation

- `any()` checks all the values.
- At least one value is `True`.
- Therefore, `any()` returns `True`.

We can remember it as:

At least one `True` → `True`
All `False` → `False`

## 7. all()

The `all()` function returns `True` only when all values in an iterable are `True`.

If even one value is `False`, it returns `False`.

### Example

In [52]:
values = [True, True, True]

result = all(values)

print(result)

True


### Explanation

- `all()` checks every value.
- All the values are `True`.
- Therefore, the result is `True`.

We can remember it as:

All `True` → `True`
At least one `False` → `False`

## 8. isinstance()

The `isinstance()` function checks whether a value belongs to a particular data type or class.

It returns either `True` or `False`.

### Example

In [53]:
age = 23

print(isinstance(age, int))
print(isinstance(age, str))

True
False


### Explanation

- `age` contains the integer value `23`.
- `isinstance(age, int)` checks whether `age` is an integer.
- It returns `True`.
- `isinstance(age, str)` checks whether `age` is a string.
- It returns `False`.

This function is useful when we need to check the type of a value.

## 9. id()

### Concept

The `id()` function returns the unique identity number of an object during its lifetime.

It can be used to identify whether two variables refer to the same object.

### Example

In [54]:
name = "Nandini"

print(id(name))

1670632100304


### Explanation

- `name` refers to a string object.
- `id(name)` returns the identity of that object.
- The exact number can be different each time the program runs.
- The `id()` function is mainly useful when working with object identity.